File: cs/Minsk/CodeAnalysis/Compilation.cs
using System.Collections.Immutable;
using Minsk.CodeAnalysis.Binding;
using Minsk.CodeAnalysis.Syntax;

namespace Minsk.CodeAnalysis;

public sealed class Compilation
{
    public Compilation? Previous { get; }
    public SyntaxTree SyntaxTree { get; }
    private BoundGlobalScope? _globalScope;

    public Compilation(SyntaxTree syntaxTree)
    {
        SyntaxTree = syntaxTree;
    }

    private Compilation(Compilation previous, SyntaxTree syntaxTree)
    {
        Previous = previous;
        SyntaxTree = syntaxTree;
    }

    internal BoundGlobalScope GlobalScope
    {
        get
        {
            if (_globalScope != null)
            {
                return _globalScope;
            }

            var globalScope = Binder.BindGlobalScope(Previous?.GlobalScope, SyntaxTree.Root);
            Interlocked.CompareExchange(ref _globalScope, globalScope, null);

            return _globalScope;
        }
    }

    public Compilation ContinueWith(SyntaxTree syntaxTree)
    {
        return new Compilation(this, syntaxTree);
    }

    public EvaluationResult Evaluate(Dictionary<VariableSymbol, object> variables)
    {
        var diagnostics = SyntaxTree.Diagnostics.Concat(GlobalScope.Diagnostics).ToImmutableArray();
        if (diagnostics.Any())
        {
            return new EvaluationResult(diagnostics, null);
        }

        var evaluator = new Evaluator(GlobalScope.Statement, variables);
        return new EvaluationResult(ImmutableArray<Diagnostic>.Empty, evaluator.Evaluate());
    }
}

File: nim/src/minsk/codeAnalysis/compilation.nim
import std/tables

import minsk/minskObject

import binding/binder
import evaluationResult
import evaluator

import syntax/syntaxNode
from syntax/parser import SyntaxTree

type
  Compilation* = ref object of SyntaxNode
    syntax*: SyntaxTree

proc newCompilation*(syntax: SyntaxTree): Compilation =
  new(result)
  result.syntax = syntax

proc evaluate*(
  node: Compilation,
  variables: TableRef[string, MinskObject]
): EvaluationResult =
  var binder = newBinder(variables)
  var diagnostics = node.syntax.diagnostics
  if diagnostics.len > 0:
    return evaluationResultError(diagnostics)
  let boundExpression = binder.bindExpression(node.syntax.root)
  diagnostics = binder.diagnostics
  if diagnostics.len > 0:
    return evaluationResultError(diagnostics)

  var evaluator = newEvaluator(boundExpression, variables)
  let value = evaluator.evaluate()
  return evaluationResultOk(value)
